### ISSUE TITLE
(Discussion) Add "Free Transform Tool"

### ISSUE THREAD

**Issue by Agnyy** [opentoonz/opentoonz#2920]
We want to add a standard tool for animating the layer. Such a tool can be found in almost all programs for animation, but for some reason there is no OT in it, another "All" tool is used instead. And therefore, what would be more convenient for everyone, we want to add the familiar tool "Transform Tool"

### Screenshot or Video Reference

![image](https://user-images.githubusercontent.com/25137502/70213455-6ce22480-174a-11ea-9709-4710c88a6e4d.png)
![image](https://user-images.githubusercontent.com/25137502/70214756-0a3e5800-174d-11ea-9e20-66acae9a8bd6.png)

In Flash 
https://www.youtube.com/watch?v=3wVoEF-HefQ

In ToonBoom
![image](https://user-images.githubusercontent.com/25137502/70213894-79b34800-174b-11ea-91f9-f66b7978f9b6.png)
https://docs.toonboom.com/help/animate-pro/Content/HAR/Stage/017_Cut-out_Animation/020_H1_Animating_using_the_Transform_Tool.html

Video in Russia 
https://www.youtube.com/watch?v=pa651-jmI0o&feature=emb_logo

**Comment by Agnyy** [comment]
![image](https://user-images.githubusercontent.com/25137502/70214211-137af500-174c-11ea-8f28-14f65d832e33.png)

Visually, the instrument will probably look like in ToonBoom. Because it’s more contrasting and pleasant to work with such a tool, or you can make it just blue, as they do now in many programs, including Photoshop.

**Comment by artisteacher** [comment]
So pretty much this is the same function as "All" mode, but in a way that may be more user friendly/expected.

**Comment by RodneyBaker** [comment]
We almost have this via the current Selector Tool.
The difference being that the Selector Tool works on parts of an image (whether vector or raster) while the Animate Tool works on entire levels.

An issue with adding this to the Animate Tool is that there is no mechanism for selecting and manipulating only parts (objects/groups) in a level.  This is why there is a separate Select Tool for that purpose.

Question:  Could the current Select Tool be presented to the user in such a way that accomplishes the goal stated here (basically by adding the Selector Tool to the bottom of the list of Animate/Edit Modes)?  Then when selected the user would simply be transitioned into working with the Selector Tool.

**Comment by RodneyBaker** [comment]
Closing but please continue the discussion in this topic...

**Comment by caryoscelus** [comment]
@RodneyBaker 

> This feature request is basically an interface for that.

While it can probably be implemented that way, i think this FR is much less ambitious. As @artisteacher points out, this would be just a different interface for the existing Animate tool functionality. I think the best way to implement it would be to reuse Selector tool code, but even if that's too much hassle it's possible to reimplement similar functionality inside All-like tool

**Comment by gab3d** [comment]
While it's always good to have options, honestly I don't see much advantages in whats proposed over what we currently have. 
Plus, as the proposed frame would enclose the whole layer limits, if the layer size exceeds the viewer limits we'd be forced to zoom out just to be able to operate on the layer, which is not currently the case. 🤔 

Nevertheless, if something like this is implemented I'd suggest it to be a Preference for the **All** transform mode, so if a user opts for having this kind of manipulator for the tool, it will always have it when selecting the All mode.

**Comment by Agnyy** [comment]
Such a tool is needed! I say this from my wide experience of using different programs, and the All tool is not as convenient as the tool we offer. This request is needed not only to me, but also to people who are switching from Flash. This tool will work with columns and has its advantages. For example, such a tool is more convenient to compress or expand the object in more proportionate manner. I like to do more accurate squash and stretch. Now it’s difficult to do this through the All tool, and it will be easier to do this for both me and new users who are used to another management.

**Comment by manongjohn** [comment]
I would also prefer a box for doing deformations similar to how the Selection Tool works, however, in regards to @gab3d's comments, this would potentially only be good for Vector levels whose bounding box would be based on the strokes in the level.  For any raster based level, the box is pointless because it would have to encompass the entire canvas size.  You couldn't select a portion of the canvas to deform.

**Comment by gab3d** [comment]
@Agnyy 
please, let's think out of the box (no pun intended)... 
you made clear that you want this kind of manipulator similar to Adobe's free transform one, also from what you say I understand that the only advantage of having it would be being able to "squash and stretch an object more conveniently and to use such a tool in a more proportional manner".
I'd like to know if there are any other functional advantages of the proposed manipulator vs the existing one, and also if you think the existing manipulator could be improved somehow to achieve the desired functionality.

Personally I don't have anything special against what you propose, but also nothing special against what is there in OT now. (Perhaps the All manipulator could be outlined to make sure it contrasts well with any possible background, as the recent changes did to the selection frame).
So if what you see as advantages could also be made part of the current manipulator it'd be good too for OT.

**Comment by Agnyy** [comment]
I do not understand the essence of the discussion. I want to introduce an additional tool, but the old tools will remain. But I do not spoil what is, I only offer one more type of tool. The tool that we are currently using in OT is not convenient and not intuitive, and during the animation it is not felt and not convenient, unlike the classical tool that I propose to introduce into OT as an additional one. If you are really interested in the benefits of such an instrument, then write to the cartoon studio and talk with specialists, for example, in Russian cartoon studios, they reacted very badly to the All tool and it seemed insanely inconvenient and complicated to them. A classic tool is much faster and more understandable.

**Comment by RodneyBaker** [comment]
@Agnyy Didn't you add "(Discussion)" to the title of this topic?

**Comment by Agnyy** [comment]
I'm really interested in hearing the opinions of others. I thought to hear a conversation about what else you can add. And now I hear only about what not to add.

**Comment by manongjohn** [comment]
@agnyy Why have 2 tools to do the same thing? At the end of the day, what you are proposing is really a different interface for the All tool.

**Comment by Agnyy** [comment]
Because these two tools are used by different people. I generally can not use the tool All, like many others from whom I asked their opinions. Therefore, I offer an additional tool. This tool will not interfere with old users, but for new users it will be easier to work. Understand the tool that now is not effective in speed. I think that it makes sense for you to ask the majority of studio animators for opinions and understand that this tool is very necessary in the studios. Since it will be easier and faster to work with him, but what is crazy insanely convenient now and spends a lot of time.

**Comment by RodneyBaker** [comment]
I can't speak to others comments but my concern is that it appears you desire the Transform Tool to manipulate objects whereas with the Animate/Edit Tool it can only transform levels.  This is what your image of the ballerina seemed to suggest as that image is a group of shapes.

As long as you understand what adding a Transform option doesn't allow for transforming anything but the level I see no reason why a developer shouldn't champion the idea.

My underlying concern is that others will think the tool is capable of more.

There are other options as well that might not be ideal.
For instance, I can readily see where a Preferences option might allow the All option to have the look and feel that you suggest and users could decide for themselves which 'widget' they prefer.

**Comment by gab3d** [comment]
@Agnyy 
easy man, I'm not saying it can't be done, just don't share your passion for the proposal.
As always, if something is added as an option and doesn't replace what is already there, it's ok for me.

Just to clarify on the terms at use, you don't want to add a new *tool*, but a new *gizmo* or *manipulator* for the Animate tool.

My suggestion is to not include it in the drop down as you showed in your mockup, because All and Transform would be exactly the same (just with a different gizmo). But instead create a new setting in the Preferences dialog to let users choose what gizmo they prefer for the All mode of the Animate tool.

**Comment by manongjohn** [comment]
The 2 tools do the same thing, they transform the level by scaling, sheering, rotating, moving, whatever...the only difference is you and others in your studio don't like the interface of the current tool.  Understandable.  

I am not a part of any studio nor do I have any studio contacts, but I know some members here are. For those of you who are part of a studio that use OT, please do chime in on the request. Obviously more insight from other studios who do use it would be helpful.

At any rate, I can see your point, and benefits for a boxed interface vs what we have now, but you don't need to be rude about it. That will not gather the support from the community nor attract a developer to make those changes.

If you are that passionate about having this done, then I suggest that you and others who feel the same put together a bounty for it, a bounty large enough to attract some developer to do it.

**Comment by Agnyy** [comment]
I do not understand. This dialogue looks like you are mocking me. I have already explained everything to you several times. Let's put it on the shelves again.
I myself am a animator and worked for the largest Russian studios and of course I communicate with animation studios around the world. I have a lot of experience in the production of animation, which studios consult with me to improve and speed up the production of cartoons, as well as optimizing time and costs. And once again I repeat this request I need as well as the STUDIO! And if you talked more often with cartoon studios, then you would know that yourself.
And now in order.

1. This tool helps you make animations faster with Tweens. And speed in production is very important! This approach is used in Cut-out animation technology.

2. Yes, this tool "transform only levels" is as it should be. They also animate in Toon Boom, only there they also group these levels. So STUDIO animators will have no problems with this, they are used to this approach. And as I said, this tool is needed for animation with the help of tweens, and not just image deformation.  https://www.youtube.com/watch?v=64XzZWbIkdo

3. Why should the tool be there? Because there it will be easier to find NEW animators! Sometimes listening to you, it seems to me that you want to make a program only for a narrow circle of people, and so that the rest could not figure it out. I propose solutions that can help new users and STUDIOS, while not harming other users. And the idea of ​​hiding this function in the settings is very strange, I ask you not to forget that it will be difficult for new users to find this tool and they just close the program and do not open it again. And the settings of GIZMO I don’t even know where are located.

So I recorded a video specifically for you. Here the practical character is fully animated only by this tool. And yes, in the studios now, using such technology, I now very often animate. Approximately 80 percent of all cartoons. There has long been practical not using frame-by-frame animation. And if I were animating with the All tool I would need a lot more time. https://www.youtube.com/watch?v=4q5ZGuuqz3w&feature=youtu.be

**Comment by Agnyy** [comment]
@manongjohn
I do not speak rudely. This is Google translator translates so. Unfortunately I do not know English properly to speak it.

At the expense of development. I already found who will do it, just a dialogue here.

**Comment by Agnyy** [comment]
In general, I would also like you to treat me with understanding. After all, I also contributed quite a lot to OT. I am constantly looking for new programmers, and this is very difficult. You say that I’m here to be an ordinary crook and I don’t understand what. It feels like you perceive only programmers and don’t notice who brings them and asks them to make functions. Sometimes I have to spend weeks, or even months, talking with programmers. Often I pay to programmers from my pocket, but for me this is often a tangible amount. And I don’t even thank anyone. This is very disappointing. I ended up having a visit sometimes and there are good programmers who agree to help Such as the  @caryoscelus  . But there are very few such good and kind programmers. For all this time I have been working with 6-7 programmers. I invested a lot of my free time in these posiki, wrote documentation often paid them. Once a programmer tricked me at all for money. I’m really offended sometimes, I’m doing something, and you treat me arrogantly. I did not come to you for help; I came with the project that I am currently doing.

**Comment by manongjohn** [comment]
I was not mocking you, no one was. . We were trying to "discuss" with you the request you made so we can see value in it and find the best approach to meet everyone's needs and concerns.

Glad you found someone to work on it already. I would suggest before adding a second command, find out if the existing command can be changed to have your boxed interface especially if it will benefit everyone. Remember, when OT was first released we didn't have an All option. It used to be there before the release I think and someone requested it brought back. Now may be a good time to change it.

Thank you for insulting me further just because I don't work with and talk to other studios, like I don't know anything and am not able to understand. By the way...there are other people who use this software, not just studios. I watch and listen to users suggestions to improve this software and have changed my initial approach to a different one even though I didn't like it or what was wanted.

For the record, you posted your idea publically which I said I agreed with.  If you post something in an open forum everyone can comment on it, including me. Expect it to be debated and criticized and be prepared not to get the feedback you expected. It happens...it had happened to me a few times and I just dealt with it.

I will stop talking about this now as it appears i am not welcome in this discussion.

**Comment by hayoko** [comment]
Why close this?

I agree with @Agnyy's proposition, anyone who has worked in Harmony will know why this is the better way to transform an object. The current All gizmo is unusable, try using it on an XP-Pen 13.3/15.6 or Cintiq 16, even with a large monitor I have trouble using it as the sensitivity is whack, there's a major lack of precision, can we not just begin to standardize parts of Toonz?

Why do I have to adjust the center anchor point in the current gizmo to the bottom just to apply a small squash? Much quicker just dragging the top center anchor from a staple bounding box.

Etc...

**Comment by RodneyBaker** [comment]
Hayden said:  

> Why close this?

As _Standard Operating Procedure_ we are still closing Feature Requests without a developer attached to the feature.  In the coming year it seems likely this will change.  There are other reasons for the closure as well such as to clear the view of works in progress as we head into v1.4.  Is there a developer attached to this feature?  Is this a WIP? Will the feature get into v1.4? At present that is hard to say especially if no dedicated developers are attached.  It is however still an ongoing discussion about the feature.  

Now perhaps we can leave off discussing the merits of the thing and actually develop the feature.  :)

**Comment by RodneyBaker** [comment]
Is there a fork where we can see the development of this feature in progress?

**Comment by gab3d** [comment]
@Agnyy 
I subscribe to @manongjohn comments.
1. nobody is mocking at anybody here.
2. when anyone proposes features/changes, he/she should be open to a mixed reception of those ideas, also know that different points of view are legitimate and not against somebody.
3. you should treat others presuming they know similar things to you and are as well intended persons as you, because you just don't know their professional background, and all of us here are giving our personal time in favor of helping develop this program in the best possible way.

So, if you think everybody is mocking at you just because the reception to your idea was less stellar than what you thought it should have been, perhaps the problem lays on your side of the line.
My advice is: expect you will always have to convince people of your ideas by telling the specific things that make them so fabulous , and don't get angry by having to do that. It's just the normal process!

Finally, all of us are giving something for free here, I salute the efforts you just told us you do to support OT and please know that your voice is well received in the community, we just ask you a little further effort in trying to improve the process of communicating your points of view with the other members in here. 
I suppose this is not much to ask, in comparison to all what you've already done for OT!

**Comment by artisteacher** [comment]
As @gab3d said, everyone here has good intentions. We all want OpenToonz to improve and hopefully be adopted by more individuals, schools, and studios. It's really easy to misinterpret the intended tone of written words, even when translation and second languages are not involved. Personally I am very grateful to everyone who contributes - whether it is money like @Agnyy or a good deal of unpaid time like @manongjohn. Most of the improvements/fixes made to OT make my life as a teacher a lot easier.

> At the expense of development. I already found who will do it, just a dialogue here.

Since there is a developer lined up, this should probably be reopened.


> My underlying concern is that others will think the tool is capable of more.

Perhaps the style could be slightly different than a standard selection box, so that there is a visual cue that the purpose is a bit different. For example, a different color from the standard selection box (though two toned for visibility on any color/value) and perhaps circles instead of squares for the corners/handles...

**Comment by caryoscelus** [comment]
@RodneyBaker there isn't yet, i've been mostly exploring relevant code bits so far (besides [fixing that bbox thing](https://github.com/opentoonz/opentoonz/pull/2854)). I'll post one when there is any actual progress, but i wanted to make sure i'm not missing anything before starting main work. Since i'm not using OT myself, i've initially missed Selector tool, so i haven't looked at that part yet. I've been wondering whether there is maybe similar relevant functionality in yet another place? Or anything else i should consider from coding standpoint? My plan so far is to try reusing Selector code if possible

**Comment by gab3d** [comment]
Specifically, I have a few questions regarding the functionality of what you are proposing:

I see two things apparently implied in your proposal about the Animate tool that are different from what is currently there:
- a temporary pivot point.
- being able to select several items to transform at the same time.

Neither of these two things are specifically related to the proposed new rectangular gizmo, but instead to the functionality of the Animate tool itself.
- As you may know an object pivot point is currently not animatable in OT (as in most animation software), so once you move it it will remain in the new position permanently. 
So I wonder how this "temporary pivot point" feature would work behind the curtains to take this into account?
- I also wonder how selecting several objects for transforming would work regarding pivots? I mean, each selected object will have its own pivot, but to be able to rotate or scale them in a grouped fashion you would have to compensate changing each object position as well? Or perhaps I misunderstood, if that's the case please correct me. 

In the mean time, as for the lack of precision of the current manipulator of the All mode, please know you can keep the **Alt** key pressed before starting to drag, to gain more precision while performing any transformation. 
Besides that, I think a bigger manipulator would help users too, and for sure there could be more things to improve in it.

**Comment by gab3d** [comment]
@caryoscelus 
[Here](https://opentoonz.readthedocs.io/en/latest/creating_movements.html#animate-tool-handle) is all the info present in the manual about the Animate tool All mode manipulator, perhaps it could be of any interest to you, as a resource of information.

**Comment by RodneyBaker** [comment]
@Agnyy I edited your topic title to show this is a Work in Progress.
Feel free to further adjust the title as you feel the need.
If I 'hide' some of the comments in this topic it's only so that the focus will remain on your proposed feature.  Hidden comments can be reopened/expanded and viewed as needed.

@caryoscelus Very glad to hear you are working this.  :)
There is surely some useful code that can be leveraged in the Selector Tool that might fit your current need.  There was recently an update to the visual aspects of the Selector Tool and that should be useful as well.

**Comment by RodneyBaker** [comment]
Perhaps too radical an idea:

It might be ideal to write this tool so well that it would be the default option for the Edit/Animate Tool.  Let me try to explain:

Background
All of the current options in the dropdown menu are 'transforms' except the last; 'All' widget.
Position, Rotation, Scale, Skew and Center (pivot/offset) are all transforms.  The 'All' option is simply an interface for manipulation of each of those transforms directly on the canvas.  In short, it is an 'All Transforms' option.

The default value of the current dropdown menu is 'Position'.
Ideally it might be this new 'Transform' state.

Consideration for new default
Note here that the proposed Transform option can and likely will still default to 'Position'.
As such the default would remain the same.  Moving/Animating something on the canvas via the new default 'Transform' option would still adjust the values in x and y channels the same as it does currently via 'Position'.  The user could then simply move the pen/cursor to see the other transform options and edit/animate accordingly.

Benefits
Improved editing/animating by default settings
No additional 'transform' option at the bottom ('transform' or 'transforms' would instead be seen at the top of the menu before the user even clicked on the dropdown menu.
Having 'transforms' showing will immediate let the user know what to expect if they open the drop down menu.
Once the drop down menu is opened the user will see the options we see today (the header would simply be 'transforms'.

There are additional considerations related to this but I'll save them for another day.

Submitted for you consideration...

**Comment by hayoko** [comment]
The code is there in the select tool already, just pipe in the transform functions.

The only thing we're missing is being able to lock the pivot position, so I could lock it to the bottom to create a squash. Center position is center unless the user drags the pivot on the transform.

![image](https://user-images.githubusercontent.com/36535462/70376804-d443d400-1904-11ea-95b9-b1bb95650442.png)

Done, we'd finally have a useful transform tool. 🤷‍♂

**Comment by Agnyy** [comment]
I would like to apologize to everyone. I looked angry, although I was not evil, I was just sad. I didn’t want to spoil anyone. I love the dialogue and support it. It's just that sometimes it’s very difficult for me to express my thoughts. And it’s hard to write one and the same thought several times to make it clearer. Therefore, I sometimes feel helpless in these dialogs when I need to defend my ideas. You probably had to draw emoticons in messages, then my messages would be nicer :-). I am very ashamed that I was able to upset someone. I’m not on purpose, I just don’t know how to write correctly. :-) I really appreciate and respect you all.

**Comment by Agnyy** [comment]
@RodneyBaker 
Unfortunately, I still can not comment on your idea because I did not understand it. I realized just that you want to add a new default function. Maybe that’s good.

![image](https://user-images.githubusercontent.com/25137502/70417292-fa3ab700-1a71-11ea-9487-c1b815bd3567.png)
@hayoko  I think in the future to add a special tool that would fix the anchor point in the right place.

**Comment by artisteacher** [comment]
> No additional 'transform' option at the bottom ('transform' or 'transforms' would instead be seen at the top of the menu before the user even clicked on the dropdown menu.


@RodneyBaker - I'm also not quite sure what you mean. How could this tool/mode be the new default of the Animate tool, yet still set to position? 

If there were a transform label, it really needs to be to the side, or displaying only when the drop down is expanded. Otherwise, if the widget is out of view (or the widget blends in too much with the background, or there's a new user who doesn't know the widgets yet), it will take more time to figure out which mode is active.

**Comment by RodneyBaker** [comment]
@artisteacher It would be much like the Selector Tool (even using much of its code as has been suggested).  The 'widget' wouldn't need to be activated because it would be the default.

As for why position would be the default for that... The Edit/Animate Tool isn't the Selection Tool, it has a default mode (currently position) and so would still retain that.  Further because the Animate Tool automatically selects all of a level (or all vectors) there is no need to select anything** so the the 'everything that is preselected can just be moved, similarly as now where the All widget can just be moved/positioned just by click dragging.

In essence the user would not sense the Transform state as a widget as it would just turn on when Edit/Animate mode is engaged.

I can understand why this is difficult to envision because Opentoonz is rather unique in how it currently cannot transform/animate objects.  It will all get sorted out and In time we'll get there.


**Obviously the goal would be to improve capabilities to Edit/Animate select (i.e. Group) and transform parts of a vector level as well but that is a much bigger task to code (REF:  Recently closed Curve Animation PR).  I'd bore everyone with talk of 'Named Groups' and exposing Groups in a Project Workspace Listing but... no few would understand that either.

**Comment by artisteacher** [comment]
@RodneyBaker - So the selection box would be visible at the same time as the various widgets in each mode. Would you be able to use it freely the whole time, or would functionality be locked to the transform mode? 

Would you see the active selection?
<img width="269" alt="Screen Shot 2019-12-09 at 10 12 30 AM" src="https://user-images.githubusercontent.com/21159116/70447419-a7aacc00-1a6c-11ea-9af7-72f9887bb9b7.png">
Or just the box?
<img width="386" alt="Screen Shot 2019-12-09 at 10 14 01 AM" src="https://user-images.githubusercontent.com/21159116/70447428-ac6f8000-1a6c-11ea-8fc6-03333fd4fd14.png">
For vector levels would the visible boundary be based on the content like it now? If so, how would you deal with potential areas of overlap between the current widget and the additional selection box?

Personally, at this point I'd rather have a preference to choose between either a "widget view" or "selection box view". Otherwise there is a potential for this to get a bit too busy visually.

On a related note,  I'd really like to have some way for the "all mode" to be the default transform mode. It could be literally set as the default. There could be a preference to choose your own default transform mode. The transform mode could be a "sticky" setting that is remembered between sessions.

**Comment by RodneyBaker** [comment]
I hesitate to answer because the foundation has not yet been laid upon which to build but...

Yes, as originally proposed manipulators would be visible that allow the various transforms. (see Ballerina image posted above)

I will guess the entirety of transform options would not appear until a mouse/pen click enabled that view however.  Then the current state of the transform could be persistent (or revert back to default) based on user preference.  In this way the user could override position as the default say with 'rotate' and not have to keep reselecting it.  (assuming they were focused on rotating things).  This also would be the moment where the state is recorded for undo purposes so that any transform change can be walked through forward/backward via undo/redo. 

> Personally, at this point I'd rather have a preference to choose between either a "widget view" or "selection box view". Otherwise there is a potential for this to get a bit too busy visually.

Don't sweat the UI too soon.  UI discussions never end but be confident that this will be resolved to general satisfaction based on optimal workflow and of course EVERYONE will be ecstatically happy with the end results.  Since we living in the dream for this feature we can write how the dreams ends.   :)

> I'd really like to have some way for the "all mode" to be the default transform mode.
Perhaps that might be a good start to this project... our first Milestone to hit!
- [ ] Add Preference to allow user to select default Transform mode in Edit/Animate Tool.

Then we can head off any objections to what gets implemented next.

**Comment by RodneyBaker** [comment]
@caryoscelus Can we confirm this feature enhancement is slated for 2020 development cycle.
This to help others understand not to expect seeing it in the v1.4 release?

**Comment by caryoscelus** [comment]
@RodneyBaker yes. I'll try to and there is a chance i will find enough time to work on it in the remaining days, but it's better to not rely on that

**Comment by RodneyBaker** [comment]
I must assume this effort is not an active work in progress so have edited the title accordingly.
@Agnyy Please feel free to further adjust the title as necessary.
